controllers/automation/policyautomation_controller.go

```diff
@@ -89,6 +89,100 @@ func (r *PolicyAutomationReconciler) setOwnerReferences(
 	return nil
 }
 
+// getTargetListMap will convert slice targetList to map for search efficiency
+func getTargetListMap(targetList []string) map[string]bool {
+	targetListMap := map[string]bool{}
+	for _, target := range targetList {
+		targetListMap[target] = true
+	}
+
+	return targetListMap
+}
+
+// getViolationContext will put the root policy information into violationContext
+// It also put the status of the non-compliance replicated policies into violationContext
+func (r *PolicyAutomationReconciler) getViolationContext(
+	policy *policyv1.Policy,
+	targetList []string,
+	policyAutomation *policyv1beta1.PolicyAutomation,
+) (policyv1beta1.ViolationContext, error) {
+	log.V(3).Info("Get the violation context from the root policy %s", policy.ObjectMeta.Name)
+
+	violationContext := policyv1beta1.ViolationContext{}
+	// 1) get the target cluster list
+	violationContext.TargetClusters = targetList
+	// 2) get the root policy name
+	violationContext.PolicyName = policy.ObjectMeta.Name
+	// 3) get the root policy namespace
+	violationContext.Namespace = policy.GetNamespace()
+	// 4) get the root policy hub cluster name
+	violationContext.HubCluster = policy.GetClusterName()
```

Comment:
Where is this defined? I'm having a hard time finding this in the diff.

Reply:
This is the function we discuss the logic yesterday.